### ISSUE TITLE
Pane resizing breaks bash, zsh, csh prompts.

### ISSUE THREAD

**Issue by ruipgpinheiro** [tmux/tmux#516]
Hello,

For some reason, the resizing feature of tmux 2.2 seems to break the bash and zsh prompt. Csh breaks as well if you type something while the prompt is line-wrapping.

I'm using xterm, and can reproduce it every time even when running the shell with "-f", by changing the prompt to something simple (it happens no matter what prompt I use), splitting the window into 4 panes, and shrinking one of the panes down to 1x1. Upon resizing that pane back to a normal size, the whole prompt is corrupted.

For example, if I set the prompt to "1234567", follow the above steps, and then type "a" to find out where the cursor is:

csh:
> 1234567a

bash:
> 1234567a
> 1234567
> 567
> 345671212
> 2123456

zsh:
>1234567
>1234567
>1234567
>1234567
>1234567
>1234567
>1234567
>1234567
>1234567
>1234567
>1234567
>1234567
>1234567a

As can be seen, the only prompt that seems to be unaffected by the issue is csh. I am unable to reproduce this issue outside tmux when using xterm directly.

When you have large prompts like I do, or even a right-sided prompt, this issue occurs *anytime* you resize and/or split a pane, and is very very annoying as it corrupts your whole buffer. I spent the last 2 weeks searching the web for solutions, trying work-arounds, but none have fixed it.

From what I've been able to gather, the issue seems to be that when zsh and bash redraw the prompt, they are not starting from the beginning of the "buffer line", but instead from the beginning of the current terminal line, which means if the line reflows part of the prompt will stay in previous lines. When resizing back to a larger size, tmux reflows the text meaning that part of the redrawn prompt will go back into previous lines, polluting the buffer even more. Interestingly, bash seems to be able to recover to the correct cursor position, while zsh does not.

Csh however "cheats", in that it does not redraw the prompt until you start typing again (both zsh and bash seem to redraw automatically on sigwinch instead), which means that by then it can clear the whole line correctly again, as it's not longer wrapping. Csh also breaks the prompt if you type something while the prompt has 1x1 size, then delete it before enlarging it back to the normal size:

>11234567a
>12
>12
>12
>12
>12341234567

As such, I am thinking it might be something about how tmux handles the line wrapping together with the VT100 commands to clear lines used by the shells when redrawing the prompt. A fix could be to have the "clear whole line" or "go to beginning of line" VT100 commands return to the start of the buffer line, and not to the start of the current on-screen line (which, in case tmux is causing the prompt to line-wrap, will not be where the shell wants to go to).

For reference, my .tmux.conf is attached:
[tmux.conf.txt](https://github.com/tmux/tmux/files/434467/tmux.conf.txt)
I have also tried using `set -g default-terminal "xterm-256color"` and `set -g default-terminal "tmux-256color"`, to no avail.

Any ideas?
Thanks,
Rui

**Comment by nicm** [comment]
tmux is at the mercy of applications running inside it, if they do not mark a line ending with an explicit newline, tmux will join the lines together when it needs to.

**Comment by ruipgpinheiro** [comment]
The problem is that this seems to happen with all shells I tried, and it's not necessarily the fault of the application. For example, the easiest case to explain is zsh. Let's also assume a simple case where the prompt itself is not multi-line:

First, zsh draws the prompt:

>PROMPT$user_command

If the user then resizes the tmux pane, zsh catches the sigwinch signal, erases the prompt by doing (I assume) ^[2K to erase the whole current line and redraws the prompt.

If the user resizes the prompt, but the text doesn't fit, it will line-wrap BEFORE zsh catches the sigwinch signal:

> PROMP
> T$use
> r_com
> mman
> d

Now zsh catches the sigwinch, erases the current line, and redraws the prompt. Notice there are no line breaks!

> PROMP
> T$use
> r_com
> mman
> PROMP
> T$use
> r_com
> mman
> d

Now we increase the size again to the original size, but due to the "prompt duplicate" everything might not be able to fit completely in a single line:

>PROMPT$user_commanPROM
>PT$user_command

Again, zsh catches the sigwinch, clears the cursor line, and redraws the prompt:

>PROMPT$user_commanPROM%
>PROMPT$user_command

This seems to be due to the tmux line-wrapping functionality, in particular the way reflow is implemented. This issue is not reproduceable outside tmux, most likely because the terminal I use (xterm) does not do reflow.

If you think the problem is with the applications themselves, I guess we should submit tickets to the specific projects themselves. But all three shells I've used in the past (csh, bash, zsh) exhibit this exact problem, so I'm starting to wonder if maybe the problem might be with tmux? Even if not, is there something I could use as a work-around?

Right now I am "fixing" it by printing a "\n" in the zsh preexec function, and having the VT100 commands to "erase line", "cursor up", "erase line" and then a "\n" at the start of my prompt. This way whenever a resize is done it deletes the line above the cursor, and redraws the prompt more or less correctly (with some extra empty lines). It does not fix the whole issue, for example if you use the mouse resizing functionality, such that you're sending multiple sigwinchs to the shell, so that it is still redrawing while you're still resizing.

**Comment by nicm** [comment]
So your problem is JUST with the line with the prompt on, the remaining
lines in the window wrap correctly?

**Comment by ruipgpinheiro** [comment]
Yes, since those are automatically redrawn by the shell and as such seem to break when tmux resizes the pane and reflows the test. Sorry if I wasn't clear enough before.

**Comment by nicm** [comment]
Well, we can't predict what is going to happen, we can only change what is on screen at the moment.

**Comment by nicm** [comment]
So it is hard to see what else we could do here.

**Comment by ruipgpinheiro** [comment]
Yes, it's not an easy thing to solve, but it seems to be caused in part by tmux re-flowing the text, and as such I think it qualifies as an "issue".

I see some possible work-arounds, but none are optimal.

For example, the "clear line" VT100 command could clear upto the previous line-break instead or to column 0. Alternatively, keep in mind which lines were just reflowed, so that the VT100 "clear line" command still clears the whole "line".

Another option I see would be to be able to signal somehow that the text should not be reflowed automatically starting from the current cursor position, and then have a way to reset that command later. This would allow us to configure the prompt not to reflow, while maintaining reflow capability for previous lines.

An easy implementation of this work-around for those bothered by this issue (like me) would be to just add a config option to disable reflow on the cursor line (or everywhere), to let the shell handle redrawing the prompt itself. This way, we could toggle reflow off while the prompt is being shown.

Hopefully someone has better ideas, since you probably know more about how the reflow is implemented than I do.

**Comment by nicm** [comment]
TBH I am not convinced this is a serious problem and I don't think it is worth adding any additional features, and we are certainly not going to change the meaning of escape sequences. If the prompt is messed up, it is as simple as pressing enter to get a new one.

These shells seem to be in the position that they want to do complex screen drawing but they don't want to use cursor-addressable mode, so there is always going to be occasional conflicts and resizing is one.

**Comment by ruipgpinheiro** [comment]
It is quite a serious issue for anyone like me with large prompts (and right-handed prompts), who resizes the panes often. Resizing the pane a few times (especially when mouse-dragging) with a large prompt can very easily lead to the situation where more than half the window is covered with a corrupted buffer.

And I don't think you'll find more widespread-use shells than csh, bash and zsh, so it's not even just "rare".

Again, I would be more than happy with a simple config to disable reflow completely (`setw -g text-relow off` or something), as I personally don't care much about that functionality. When I'm not using tmux, I use xterm directly: xterm doesn't reflow text at all (when you shrink the window it just "chops" the buffer), and I never actually needed it to.

I agree that more complex fixes might be overkill (especially changing meaning of VT100 signals), but it's hard to deny that the reflow functionality is not breaking the shells.

Thanks,
Rui

**Comment by nicm** [comment]
tmux has way too many options already and my policy is to avoid more
trivial options that are only going to be used by a very small number of
people. I want to remove options, not add them.

So it needs a solution that is generally useful or that can be
implemented by default.

**Comment by ruipgpinheiro** [comment]
Right, I can understand that, but it's still a bit disappointing, as this is "extra functionality" I don't need breaks my prompt quite often (anytime I shrink a pane''s size, which is quite often). I've been actually considering not using tmux at all, and installing a tiled display manager instead on which I can use xterm directly (since it does not have reflow capabilities).

What would you suggest the prompts do different? If I knew what combination of escape sequences (or other commands) you think could be used to fix this (you mentioned cursor-addressable mode?), I might be able to work-around this issue entirely by doing it every time the prompt is redrawn.

**Comment by nicm** [comment]
The problem would be that cursor addressable mode also turns off history.

Can you run your shell in script(1) then press enter a few times so it
draws a few prompts (don't resize) then exit and send me the typescript
file so I can see what it is drawing?

**Comment by ruipgpinheiro** [comment]
I ran three tests of zsh in script.

First, [custom_prompt.txt](https://github.com/tmux/tmux/files/438997/custom_prompt.txt) is when I ran script over zsh with my custom prompt, pressed enter a few times, and exited. For reference, my custom prompt is:

> PROMPT="%{$fg_no_bold[yellow]%}%~ %{$reset_color%}%# "
> RPROMPT="%{%K{green}$fg_bold[black]%}\$(ws_prompt)%{$reset_color%} %{%F{13}%}%T %{%f%k%}"


Second, [no_custom_prompt.txt](https://github.com/tmux/tmux/files/439000/no_custom_prompt.txt) is when I started script, then ran "zsh -f" to clear the configuration, pressed enter a few times, and then exited twice (so the first and last lines are my custom prompt, but the rest are the default zsh prompt).

Lastly, [split.txt](https://github.com/tmux/tmux/files/439002/split.txt) is where, using my custom prompt, I split the pane in half, and then exit. An image of the resulting buffer can be seen in ![zsh_split_prompt](https://cloud.githubusercontent.com/assets/11169970/18001892/efc75b90-6b7c-11e6-8b24-11c1c7d12a92.png), basically the split causes the RPROMPT to reflow, resulting in a double prompt.

Hopefully these should cover most of the main cases (normal prompt drawing with default prompt; custom prompt; and redrawing after a sigwinch).

**Comment by ruipgpinheiro** [comment]
I managed to disable reflow by commenting out

    if (reflow)
        screen_reflow(s, sx);

in the "screen_resize" function in screen.c.

This seems to successfully disable the reflow functionality completely, and the issue is no longer reproduceable, which confirms that this is a tmux issue (and not a shell issue).

I hope the issue will be fixed sometime in the future (although I don't see any easy fix), but for now I am perfectly happy with this work-around.

Ideally, this would be set by a configuration variable (I could even submit a PR myself), but since you don't seem to want to add extra configurations only a few people will use, I'm happy enough to build tmux from source with those lines commented out to work around the problem, at least for now.

Thank you,
Rui

**Comment by ruipgpinheiro** [comment]
I had an idea yesterday regarding a way to fix the biggest causes of corruption in shells that react to sigwinch, which is to simply delay reporting sigwinch to shells until the size has been constant for a small time period.

As such, I changed server_client.c server_client_check_resize_event:

    /* Check if pane should be resized. */
    void server_client_check_resize_event(evutil_socket_t fd, short what, void *arg)
    {
        struct window_pane *wp = (struct window_pane*)arg;
        struct winsize	ws;
        
        if (!(wp->flags & PANE_RESIZE))
            return;
        
        memset(&ws, 0, sizeof ws);
        ws.ws_col = wp->sx;
        ws.ws_row = wp->sy;
        
        if (ioctl(wp->fd, TIOCSWINSZ, &ws) == -1) {
    #ifdef __sun
            /*
             * Some versions of Solaris apparently can return an error when
             * resizing; don't know why this happens, can't reproduce on
             * other platforms and ignoring it doesn't seem to cause any
             * issues.
             */
            if (errno != EINVAL && errno != ENXIO)
    #endif
               fatal("ioctl failed");
        }
    
        wp->flags &= ~PANE_RESIZE;
    }
    
    void
    server_client_check_resize(struct window_pane *wp)
    { 
        if (!(wp->flags & PANE_RESIZE))
            return;
        
        if (event_initialized(&(wp->ev_check_resize)))
            event_del(&(wp->ev_check_resize));
        
        event_set(&(wp->ev_check_resize), 0, EV_TIMEOUT, server_client_check_resize_event, (void*)wp);
        
        struct timeval tv = { 0, 250000 }; // wait 0.25 seconds
        event_add(&(wp->ev_check_resize), &tv);
    }

And added

    struct event ev_check_resize;

to the window_pane structure definition.

These changes make tmux wait for at least 0.25 seconds of a pane not being resized before it reports the size change to the terminal itself (causing a sigwinch), which means zsh will not reflow while the size is still changing (for example, when mouse-resizing, or when holding down a key combination mapped to pane-resize).

Together with my previous "disable reflow" fix, this has completely fix the issue reported. Note that I didn't test with reflow still enabled, but I assume it should reduce buffer corruption a lot even when using reflow.

I'm not really familiar with the tmux source structure (nor with libevent), so there's probably a better way of doing this, so take the code above with a grain of salt. Still, it works, and no more buffer corruption when resizing!

**Comment by nicm** [comment]
Why is this necessary? tmux should already be resizing the window before it sends SIGWINCH, what are these shells doing that they need a delay?

**Comment by ruipgpinheiro** [comment]
Zsh and bash resize the prompt when they receive a sigwinch (csh is different, it waits for a keypress before redrawing the prompt after a sigwinch).

However, image you are mouse-resizing the pane to be less wide. Every time you drag "1 character", it reduces the window size by 1 character, and sends a sigwinch. If you do that fast enough, you might reduce it "1 character" further while the shell is still processing the previous sigwinch.

Let's imagine you start with a width of 80. You now hold down a keybind that resizes by 1 per step, or drag your mouse until you have a width of 78. Tmux will send two sigwinches, one notifying that the size dropped to 79, and a second that the size dropped to 78. If the shell is still resizing its prompt due to the "79 sigwinch" when tmux reduces the window size to 78, the shell expects a width 1 character wider, and as such the last character will wrap to the following line. When the shell catches the following sigwinch for 78, it redraws the prompt, but since the line wrapped, this results in a duplicate prompt.

Now instead of doing 80->78, imagine doing a quick 80->40 resize. If you get the drag speed just right, you might get 40 duplicate prompts!

Waiting for 0.25s means that the prompt will only be redrawn once the size is "stable", which avoids this issue entirely (with no loss of functionality, IMO).

**Comment by nicm** [comment]
Did you try just doing it on mouse release?

**Comment by nicm** [comment]
If that doesn't work, a timer could be OK, but I think it should redraw at least every 0.25 seconds, so if the timer is already running it should not reschedule it.

**Comment by ruipgpinheiro** [comment]
No, I didn't try that. I looked at the code but couldn't find a place where I could hook the mouse drag/release event. It would also be necessary to hook "keybind release" as I can reproduce it with

    bind -n M-y resize-pane -L 2

And then holding Alt-y for a few seconds.

As for the redraw every 0.25 seconds (instead of resetting): I didn't test it, but I think that would still cause issues (although not as many), just less. Imagine you're resizing (mouse-dragging) for a full second, you'll still get 3 sigwinches if the redraw happens every 0.25 seconds, which has the possibility of causing 3 duplicate prompts with the correct timing.

Maybe you have a better idea of how to solve the issue without the timer? I mean, all I'm using the timer for is as a way of "detecting" when the resize is finished (when a resize is done over multiple steps, either due to drag-resizing or by calling resize-pane multiple times in less than 0.25 seconds).

**Comment by nicm** [comment]
Give a diff please and I'll try it.

**Comment by ruipgpinheiro** [comment]
Okay. I'm not very familiar with diffs, so hopefully I did this correctly.

[prompt_resize_workaround.zip](https://github.com/tmux/tmux/files/462251/prompt_resize_workaround.zip)
This zip file contains three diffs. The screen.c change disables reflow. The other two changes make it wait 0.25 seconds before notifying the application of a size change.

**Comment by nicm** [comment]
I like the delay, but I think we need to give applications some chance to redraw or it looks strange. It is probably OK only to do that in the alt screen (CUP). Please try this: [r.diff.txt](https://github.com/tmux/tmux/files/462416/r.diff.txt)

**Comment by nicm** [comment]
Wait actually, I don't like that there is a delay when doing eg C-b Space. I think the first redraw should happen immediately, only subsequent ones should be delayed.

**Comment by nicm** [comment]
Please try this instead:  [r2.diff.txt](https://github.com/tmux/tmux/files/462436/r2.diff.txt)

**Comment by ruipgpinheiro** [comment]
> Wait actually, I don't like that there is a delay when doing eg C-b Space. I think the first redraw should happen immediately, only subsequent ones should be delayed.

Right, I agree that commands that are "immediate" shouldn't create a delay. The delay is only necessary for things that might occur multiple times in a row, e.g. "resize-pane" and "mouse resizing". Everything else does not require a delay.

Doing the first redraw always could still result in duplicate prompts, though. If there was a way of flagging "pane in being resized right now, user still hasn't let go of mouse or keyboard combination" we could skip the sigwinch then. Alternatively, have a way to signal "resize happened because of mouse or resize-pane command" and use the timer only for those cases.

> It is probably OK only to do that in the alt screen (CUP)

I'm not very familiar with the code-base, what do you mean by this? What is the alternate screen? Why wouldn't the delay make sense for it?

**Comment by ruipgpinheiro** [comment]
Just built your latest diff (r2.diff.txt) and did some testing (using the tmux 2.2 release as a base)

With reflow disabled, I am unable to reproduce any prompt corruption at all. So effectively, 

With reflow enabled, I can still reproduce the first issue (dual prompts due to reflow), but much less than ever before.

**Comment by ruipgpinheiro** [comment]
I was thinking about a fix for the reflow issue:

Would it be possible to have reflow only apply AFTER sigwinch? Basically, let the shell resize the prompt, THEN (and only then) reflow whatever text might be left? This would guarantee that the reflow would not touch the prompt, I think?

From a quick look at the code-base, this might be a complex change, though.

**Comment by gpakosz** [comment]
Hello,

While I'm not facing the exact problem discusses in this issue, I'm might be facing something related on Darwin.

Take tmux `1.8` and follow the following steps:
  - launch `$tmux -f /dev/null`
  - sudo -i
  - *[enter your password]*
  - launch `vim` (or anything else that might change window title provided `automatic-rename is `on`)

In tmux `1.8`, renaming the window title to `vim` is instant.

Now, take tmux `HEAD` and apply #529. Repeat the steps mentioned above. While #529 fixes automatic title renaming, it's not instant anymore. I either have to wait for `status-inverval` or press e.g. Enter to get a new line/prompt displayed for the window title to get updated.

![tmux-516 mov](https://cloud.githubusercontent.com/assets/553208/18411327/844bff10-7775-11e6-9480-c0a9054d15b7.gif)

Is this related? If not, feel free to ignore me (or should I open another issue?).

**Comment by nicm** [comment]
We could reflow after TIOCSWINSZ but we do not know when the SIGWINCH
has been delivered, let alone when the process has received it and
finished whatever it wants to do with it.

**Comment by nicm** [comment]
I have applied this change now.

**Comment by ruipgpinheiro** [comment]
That's great, thanks! This definitely reduces the degree to which resizing corrupts the prompt.
Which commit? I might be doing something wrong, but I cannot seem to find this change on the commit history for the project.

Still one issue left to solve (the reflow one I mentioned before), but I've had no problems the last few weeks with this fix and reflow disabled, so lacking any other solutions, it might be good to consider adding reflow-enabled as a setting just because of this use-case.

**Comment by nicm** [comment]
You need to wait until GitHub syncs from OpenBSD which is hourly IIRC.

**Comment by amerlyq** [comment]
Actually, I'm not satisfied with this partial solution. 
I don't like delay-based compromises at all, which even didn't fixed the problem completely ;) 
Having same problem as issue reporter, same frustration and irritation through years, there simply is no way to resign myself to suck half-baked solution.
I use tiling wm, therefore my terminals are being created and resized *almost all the time*.
All neighbouring tmux instances simultaneously generate prompt crap everywhere...
I see crap each 1 minute, >500 times a day, >180k a year. 
It's at least 180k of burnt nerve cells.

I insist on the solution, similar to the first one, introduced in the thread beginning.
`'clear line' _must_ clear whole last line`.
It's completely according to VT100 specs.
But VT100 had no history. History -- is feature emulated by `tmux`.
Also VT100 couldn't be resized. Resizing -- is feature provided by `tmux`.
So, `tmux` not *only* emulates VT100, actually `tmux` *extends* its behavior.
Ability to resize and to track history inside `tmux` requires `virtual buffer` == scrollbuffer, which is interpreted and displayed (squeezed) inside actual  terminal.
For external terminal -- `tmux` is the only launched app, working in position-based mode.
But for each inner CLI application `tmux` *IS* terminal.
Each cli application works with lines of terminal.
But these lines are virtual lines of tmux -- therefore lines of scrollback!
These lines aren't lines of enclosing external terminal emulator.
Even if you map single long `tmux` virtual scrollbuffer row onto multiple short rows of external terminal -- for CLI app it continues to be single long row.
So, command `clear line` must clear whole last row of scrollback!
It's consistent. 
CLI app asks to clear line till `column 0`, and you clear it till column 0 `in scrollback`.
Because tmux is terminal.

I really don't understand why you would want to clear line till `column 0` of external terminal emulator, which even can be different at the same time or be disconnected completely.

**Comment by ruipgpinheiro** [comment]
@amerlyq I agree, the partial solution does not resolve the issue as it only fixes one of two problems which caused it.

However, with this solution, disabling reflow *does* resolve the issue.

I proposed adding a setting to disable reflow, but the proposal was rejected, so I've been building my own binaries of tmux with a simple change which disables reflow:
`sed -i -e "s~if (reflow)~if (0)~" screen.c`

**Comment by amerlyq** [comment]
My wish to continue discussion on this issue originates from the simple assumption: rejection was biased and superficial.
Solution with `clear line` won't violate VT100 specifications, even if it seems so on first glance.
It's precisely the reason why I put all those details -- if nothing was violated, then nothing is over.   
Because current problem may be described as *adhering to misinterpreted VT100 specification* resulting in *suboptimal architectural/behavioral decision*.

Otherwise we need clear reasons with real usecases where clearing line till visual `column 0` of external terminal will be mandatory or at least useful.

**Comment by amerlyq** [comment]
@ruipgpinheiro, please reopen issue if you somewhat agree with my arguments.
Lets keep it open at least until @nicm feedback -- for issue to not being lost among other closed threads.

**Comment by ruipgpinheiro** [comment]
@amerlyq Sorry, I am don't seem to be able to re-open the issue.

**Comment by jiixyj** [comment]
I've pushed another attempt/hack at properly solving this here: https://github.com/jiixyj/tmux/commit/d18693f728cc037883d496a3487bd34caf2d5c92
Should I open a new issue or is reusing this one OK?

The main differences to the current approach are:
 - TIOCSWINSZ ioctls and therefore SIGWINCHs to subprocesses are not rate limited any more.
 - Instead, the reflowing step is limited and happens a few milliseconds _after_ SIGWINCH has been sent to the subprocess. This gives the subprocess a chance to redraw itself.
 - Now, the reflowing won't simply reflow the whole screen any more. It just reflows up until the line that that subprocess has drawn to between the SIGWINCH and the reflow (if it has drawn to the screen at all). The assumption here is that this line is still 'live' and under control of the subprocess, so reflowing it would result in glitches.

It works great with:
 - one line zsh, bash and csh prompts (csh cursor position won't update automatically though)
 - ncurses programs using filter(). This is a special ncurses mode that uses just one line without switching to the alternate screen. With the current approach resizing/reflowing breaks completely here. I'll try to code up a quick example.

Two line zsh prompts still have some problems/glitches though... I'm looking into it.
Do you think this is a promising approach? I don't think there are too many downsides, but I haven't looked hard enough yet.

**Comment by nicm** [comment]
Open a new PR please.

On 29 Nov 2017 6:55 am, "jiixyj" <notifications@github.com> wrote:

I've pushed another attempt/hack at properly solving this here: jiixyj/tmux@
d18693f
<https://github.com/jiixyj/tmux/commit/d18693f728cc037883d496a3487bd34caf2d5c92>
Should I open a new issue or is reusing this one OK?

The main differences to the current approach are:

   - TIOCSWINSZ ioctls and therefore SIGWINCHs to subprocesses are not rate
   limited any more.
   - Instead, the reflowing step is limited and happens a few milliseconds
   *after* SIGWINCH has been sent to the subprocess. This gives the
   subprocess a chance to redraw itself.
   - Now, the reflowing won't simply reflow the whole screen any more. It
   just reflows up until the line that that subprocess has drawn to between
   the SIGWINCH and the reflow (if it has drawn to the screen at all). The
   assumption here is that this line is still 'live' and under control of the
   subprocess, so reflowing it would result in glitches.

It works great with:

   - one line zsh, bash and csh prompts (csh cursor position won't update
   automatically though)
   - ncurses programs using filter(). This is a special ncurses mode that
   uses just one line without switching to the alternate screen. With the
   current approach resizing/reflowing breaks completely here. I'll try to
   code up a quick example.

Two line zsh prompts still have some problems/glitches though... I'm
looking into it.
Do you think this is a promising approach? I don't think there are too many
downsides, but I haven't looked hard enough yet.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/tmux/tmux/issues/516#issuecomment-347769975>, or mute
the thread

**Comment by ruipgpinheiro** [comment]
@nicm Is there any way for you to reconsider adding a way to disable reflow? I've been building tmux from source since this pull request was closed with reflow disabled and I've never once needed it.

Over the last 2 years I have encountered various other people who had similar problems and wished to disable reflow. At least one even gave up on tmux for alternatives exactly because of reflow.

Seeing as disabling reflow is a single-line change (`sed -i -e "s~if (reflow)~if (0)~" screen.c`) it does not seem like much more maintenance work to add a configuration variable.
